From b64ea99592352e1468ba9f51019987858babaff3 Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Wed, 9 Aug 2023 23:29:33 +0600
Subject: [PATCH 90/92] drivers: net: phy: marvell_88q3344_es: update probe
 code

The engineering sample of marvell_88q3344_es has id register set to zero
in all MMDs.

Implement custom match_phy_dev method that returns true if and only if
the phy is c45 and all MMDs have zero id register. This method does not
execute any device communication (and this is safe to call for arbitrary
PHY), instead it depends on phydev->is_c45 set to true and
phydev->c45_ids.device_ids[] populated at PHY device register time.
That happens if "compatible" property of the PHY's device tree node is
set to "ethernet-phy-ieee802.3-c45".

Firmware version check is kept in probe method that is only called if
match_phy_dev returned true.

Also remove any overwriting of is_c45 flag in driver methods.

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
---
 drivers/net/phy/marvell_88q3344_es.c | 45 +++++++++++++---------------
 1 file changed, 21 insertions(+), 24 deletions(-)

diff --git a/drivers/net/phy/marvell_88q3344_es.c b/drivers/net/phy/marvell_88q3344_es.c
index 69ac405f34d0..6e91875044f3 100644
--- a/drivers/net/phy/marvell_88q3344_es.c
+++ b/drivers/net/phy/marvell_88q3344_es.c
@@ -61,7 +61,7 @@ static int MRVL_AMG_MB_getStatus(struct phy_device *phydev) {
 		val = phy_read_mmd(phydev, 0x1E, 0x47);
 		//device->readReg(device, device->phyAddr, 0x1E, 0x47, &val);
 		if ((val & 0x2) == 0x2) {
-			mrvl_err("MRVL_AMG_MB_getStatus error: transaction error");
+			mrvl_warn("MRVL_AMG_MB_getStatus error: transaction error");
 			return -1;
 		}
 		if ((val & 0x1) == 0) {
@@ -307,23 +307,32 @@ static MRVL_APHY_STATUS MRVL_AMG_GetFirmwareInfo(MRVL_DEV_PTR device, MRVL_AMG_F
 }
 */
 
-
-static int mv88q3344_es_probe(struct phy_device *phydev)
+static int mv88q3344_es_match_phy_dev(struct phy_device *phydev)
 {
-	u32 regData = 0;
+	int i;
 
-	phydev->priv = &log_level;
+	if (!phydev->is_c45)
+		return false;
 
+	for (i = 1; i < ARRAY_SIZE(phydev->c45_ids.device_ids); i++) {
+		if (phydev->c45_ids.device_ids[i] == 0xffffffff)
+			continue;
 
-	mrvl_info("mv88q3344_es_probe(): phydev->port: %i  mdio-bus-name: '%s' mdio-bus-id '%s'\n", phydev->port, phydev->mdio.bus->name, phydev->mdio.bus->id);
+		/* For engineering sample, all ids are zeroes */
+		if (phydev->c45_ids.device_ids[i] != 0)
+			return false;
+	}
 
+	return true;
+}
 
+static int mv88q3344_es_probe(struct phy_device *phydev)
+{
+	u32 regData = 0;
 
-	if(!phydev->is_c45) {
-		mrvl_warn("mv88q3344_es_probe() Not c45 !\n");
-			phydev->is_c45 = 1;
-		}
+	phydev->priv = &log_level;
 
+	mrvl_info("mv88q3344_es_probe(): phydev->port: %i  mdio-bus-name: '%s' mdio-bus-id '%s'\n", phydev->port, phydev->mdio.bus->name, phydev->mdio.bus->id);
 
 	MRVL_AMG_Virtual_readReg32(phydev, 0x1F8000, &regData);
 //[   18.872127] fwInfo 0.4.1.7
@@ -335,18 +344,12 @@ static int mv88q3344_es_probe(struct phy_device *phydev)
 			(regData >> 28) & 0xF,
 			(regData >> 16) & 0xFFF,  (regData >> 8) & 0xFF, regData & 0xFF);
 
-
 	if(regData != MRVL_88Q3344_ES_FW_VER) {
-		mrvl_warn("Found PHY ID (0x%.8x) but firmware didn't match (0x%.8x)",
-					phydev->drv->phy_id, regData);
+		mrvl_warn("Unexpected firmware (0x%.8x)", regData);
 		return -ENODEV;
 	}
-	else {
-		mrvl_notice("Matched MV88Q3344 engineering sample PHY (0x%.8x)", regData);
-	}
 
 	return 0;
-
 }
 
 static void mv88q3344_es_remove(struct phy_device *phydev)
@@ -363,7 +366,6 @@ static int mv88q3344_es_suspend(struct phy_device *phydev)
 
 static int mv88q3344_es_resume(struct phy_device *phydev)
 {
-
 	mrvl_info("mv88q3344_es_resume()\n");
 	return 0;
 }
@@ -371,12 +373,6 @@ static int mv88q3344_es_resume(struct phy_device *phydev)
 static int mv88q3344_es_config_init(struct phy_device *phydev)
 {
 	mrvl_info("mv88q3344_es_config_init()\n");
-
-	if(!phydev->is_c45) {
-			pr_info("mv88q3344_es_config_init() Not c45 !\n");
-			phydev->is_c45 = 1;
-		}
-
 	return 0;
 }
 
@@ -545,6 +541,7 @@ static struct phy_driver mv88q3344_es_drivers[] = {
 		.name = "mv88q3344_es",
 		.soft_reset = mv88q3344_es_reset,
 		.probe = mv88q3344_es_probe,
+		.match_phy_device = mv88q3344_es_match_phy_dev,
 		.suspend = mv88q3344_es_suspend,
 		.resume = mv88q3344_es_resume,
 		.config_init = mv88q3344_es_config_init,
-- 
2.30.2

